src/main/scala/com/ibm/m4d/mover/conf/VaultSecretSubstitutor.scala

```diff
@@ -0,0 +1,155 @@
+/**
+  * (C) Copyright IBM Corporation 2020.
+  *
+  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+  * the License. You may obtain a copy of the License at
+  *
+  * http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+  * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+  * specific language governing permissions and limitations under the License.
+  */
+package com.ibm.m4d.mover.conf
+
+import com.ibm.m4d.mover.ConfigException
+import com.ibm.m4d.mover.conf.SecretProviderSubstitutor.SecretSubstitutionKey
+import com.typesafe.config.{Config, ConfigFactory}
+import okhttp3._
+import org.apache.commons.io.FileUtils
+import org.slf4j.LoggerFactory
+
+import java.io.File
+import java.nio.charset.Charset
+import scala.collection.JavaConverters._
+import scala.util.{Failure, Success, Try}
+
+/**
+  * This substitutor retrieves credentials from Vault using a JWT token as authentication.
+  */
+class VaultSecretSubstitutor extends CredentialSubstitutor {
+  import VaultSecretSubstitutor._
+
+  override def substitutionKey: String = VaultSecretSubstitutor.VaultConfigKey
+
+  override def substituteConfig(config: Config): Config = {
+    val vaultConfig = config.getConfig(substitutionKey)
+
+    val newConf = VaultClient.readData(vaultConfig)
+
+    val newKeys = newConf.entrySet().asScala.map(_.getKey).toSeq
+    if (newKeys.nonEmpty) {
+      logger.info("Found credentials with keys: " + newKeys.mkString(","))
+    } else {
+      logger.warn("No entries found in " + config.getString(SecretSubstitutionKey))
+    }
+    for (newKey <- newKeys) {
+      if (config.hasPath(newKey)) {
+        logger.warn(s"Key '$newKey' already exists and will not be substituted!")
+      }
+    }
+
+    config.withoutPath(VaultSecretSubstitutor.VaultConfigKey)
+      .withFallback(newConf)
```

Comment:
At the end of this all the keys found in the Vault that corresponds to the configuration passed are added to that configuration, existing keys are not replaced so substitute is a bit odd.

Reply:
I think more comments would help.

Reply:
Ok I see what you mean. The substitute refers to the substitution of the field `vault` with the values that are found in Vault. The field `vault` is afterwards removed. I see how the behaviour of not replacing existing fields is misleading. My reasoning was that whatever the user explicitly specifies (in the config) should take precedence over what is fetched from external resources such as Vault.